system/lib/libc/musl/src/thread/__wait.c

```diff
@@ -8,32 +8,50 @@
 void __wait(volatile int *addr, volatile int *waiters, int val, int priv)
 {
 	int spins=100;
+#ifndef __EMSCRIPTEN__
 	if (priv) priv = FUTEX_PRIVATE;
+#endif
 	while (spins-- && (!waiters || !*waiters)) {
 		if (*addr==val) a_spin();
 		else return;
 	}
 	if (waiters) a_inc(waiters);
 #ifdef __EMSCRIPTEN__
-	int is_runtime_thread = emscripten_is_main_runtime_thread();
+	pthread_t self = __pthread_self();
+	const int is_runtime_thread = emscripten_is_main_runtime_thread();
+
+	// Main runtime thread may need to run proxied calls, so sleep in very small slices to be responsive.
+#ifdef __EMSCRIPTEN_ATOMIC_BUILTINS__
+	const int64_t maxNsecsToSleep = is_runtime_thread ? 1000000 : 100000000;
+#else // !defined(__EMSCRIPTEN_ATOMIC_BUILTINS__)
+	const double maxMsecsToSleep = is_runtime_thread ? 1 : 100;
+#endif // defined(__EMSCRIPTEN_ATOMIC_BUILTINS__)
+
 	while (*addr==val) {
-		if (is_runtime_thread || pthread_self()->cancelasync == PTHREAD_CANCEL_ASYNCHRONOUS) {
-			// Must wait in slices in case this thread is cancelled in between.
+		if (is_runtime_thread || self->cancelasync) {
 			int e;
 			do {
-				if (pthread_self()->cancel) {
+				if (self->cancel) {
 					if (waiters) a_dec(waiters);
 					return;
 				}
+				// Must wait in slices in case this thread is cancelled in between.
+#ifdef __EMSCRIPTEN_ATOMIC_BUILTINS__
+				e = __builtin_wasm_memory_atomic_wait32((int*)addr, val, maxNsecsToSleep);
+				if (e == 2) e = -ETIMEDOUT;
+#else // !defined(__EMSCRIPTEN_ATOMIC_BUILTINS__)
+				e = emscripten_futex_wait((void*)addr, val, maxMsecsToSleep);
+#endif // defined(__EMSCRIPTEN_ATOMIC_BUILTINS__)
 				// Assist other threads by executing proxied operations that are effectively singlethreaded.
 				if (is_runtime_thread) emscripten_main_thread_process_queued_calls();
-				// Main thread waits in _very_ small slices so that it stays responsive to assist proxied
-				// pthread calls.
-				e = emscripten_futex_wait((void*)addr, val, is_runtime_thread ? 1 : 100);
-			} while(e == -ETIMEDOUT);
+			} while (e == -ETIMEDOUT);
 		} else {
 			// Can wait in one go.
+#ifdef __EMSCRIPTEN_ATOMIC_BUILTINS__
+			__builtin_wasm_memory_atomic_wait32((int*)addr, val, -1);
+#else // !defined(__EMSCRIPTEN_ATOMIC_BUILTINS__)
 			emscripten_futex_wait((void*)addr, val, INFINITY);
```

Comment:
But doesn't `emscripten_futex_wait` already used `__builtin_wasm_memory_atomic_wait32` in places where this is legal (i.e. off the main web thread?)

I think the decision on whether we can use the atomic wait needs to be dynamic doesn't it ?  We don't actually want to produce two different binaries do we?

Reply:
`emscripten_futex_wait` is implemented in JS and calls directly `Atomics.wait` for non-web environments. So, this PR is NFC in that respect.

This does indeed produce a different binary if you only target Node, so that would be a breaking change. Perhaps it would be better to implement `emscripten_futex_wait` dynamically in native code. But as far a I know, that is currently not possible since a JS equivalent of `ENVIRONMENT_IS_WEB` is not available (PR #15659 is also relevant here).

Reply:
I think we can/should implement something like `emscripten_thread_can_block` in native code.

Reply:
I decided to have a go adding`emscripten_thread_can_block` along with native version of `emscripten_futex_wait`: https://github.com/emscripten-core/emscripten/pull/15742

Reply:
Great! Do you plan to move `emscripten_futex_wake` to native code as well (which uses the `__builtin_wasm_memory_atomic_notify` builtin)?

Reply:
Yes I hope we can move all the wait/wait to native code but I'm doing it one step at a time,